dpd-client/src/lib.rs

```diff
@@ -28,3 +29,106 @@ pub struct ClientState {
     /// Used for logging requests and responses.
     pub log: Logger,
 }
+
+impl Client {
+    /// Ensure that a NAT entry exists, overwriting a previous conflicting entry if
+    /// applicable.
+    #[allow(clippy::too_many_arguments)]
+    pub async fn ensure_nat_entry(
+        &self,
+        log: &Logger,
+        target_ip: ipnetwork::IpNetwork,
+        target_mac: types::MacAddr,
+        target_first_port: u16,
+        target_last_port: u16,
+        target_vni: u32,
+        sled_ip_address: &std::net::Ipv6Addr,
+    ) -> Result<(), progenitor_client::Error<types::Error>> {
+        let existing_nat = match target_ip {
+            ipnetwork::IpNetwork::V4(network) => {
+                self.nat_ipv4_get(&network.ip(), target_first_port).await
+            }
+            ipnetwork::IpNetwork::V6(network) => {
+                self.nat_ipv6_get(&network.ip(), target_first_port).await
+            }
+        };
+
+        // If a NAT entry already exists, but has the wrong internal
+        // IP address, delete the old entry before continuing (the
+        // DPD entry-creation API won't replace an existing entry).
+        // If the entry exists and has the right internal IP, there's
+        // no more work to do for this external IP.
+        match existing_nat {
+            Ok(existing) => {
+                let existing = existing.into_inner();
+                if existing.internal_ip != *sled_ip_address {
+                    info!(log, "deleting old nat entry";
+                      "target_ip" => ?target_ip);
+
+                    match target_ip {
+                        ipnetwork::IpNetwork::V4(network) => {
+                            self.nat_ipv4_delete(
+                                &network.ip(),
+                                target_first_port,
+                            )
+                            .await
+                        }
+                        ipnetwork::IpNetwork::V6(network) => {
+                            self.nat_ipv6_delete(
+                                &network.ip(),
+                                target_first_port,
+                            )
+                            .await
+                        }
+                    }?;
+                } else {
+                    info!(log,
+                      "nat entry with expected internal ip exists";
```

Comment:
Looks like this was basically existing code, but why is comparing just the internal ip enough? @internet-diglett might have some more context

Reply:
Note it checks the response of `nat_ipv[46]_get(&network.ip(), target_first_port)` which includes the first port of the range in its query

Reply:
This one's on me. This code used to live in the instance creation saga and checked only for the existence of a NAT entry before continuing. In #3127, I extracted that out to a common routine to use both here and when a migration completes. In the latter case we need to update incorrect existing entries, which led to the creation of this code. When @jmpesp and I were testing that change we discovered that the code couldn't just call `nat_ipv?_create` unconditionally if an entry already existed. I think I added this check to try to avoid needlessly deleting and recreating identical entries, but it's entirely possible that we need to refine it.

@internet-diglett would love to get your feedback here--mea culpa for not checking with you on this before merging #3127.

Reply:
Apologies for the delayed response. For some reason Github didn't send me an email, or I overlooked it? If I recall correctly, the internal ip address and the first port are the keys used to lookup nat entries in dpd:

https://github.com/oxidecomputer/dendrite/blob/60a7c97d08d9fc26cda1e133f7f25e65895fe167/dpd/src/nat.rs#L371-L383

As for feedback on the currently implemented process of updating NAT entries, admittedly I'm still learning more about how Nexus works under the hood, but as long as we're not at risk of a race condition with regards to asic updates we should be ok (your answer to @rcgoodfellow basically satisfies any concerns I had).